src/video_core/control/scheduler.cpp

```diff
@@ -1,32 +1,204 @@
+        }
+    } while (pending_work);
+}
+
+bool Scheduler::ScheduleLevel(std::list<size_t>& queue) {
+    bool found_anything = false;
+    size_t min_schedule_count = std::numeric_limits<size_t>::max();
+    for (auto id : queue) {
+        auto& fifo = impl->gpfifos[id];
+        std::scoped_lock lk2(fifo.guard);
+        if (!fifo.pending_work.empty() || fifo.is_running) {
+            if (fifo.scheduled_count > min_schedule_count) {
+                continue;
+            }
+            if (fifo.scheduled_count < fifo.yield_count) {
+                fifo.scheduled_count++;
+                continue;
+            }
+            min_schedule_count = fifo.scheduled_count;
+            impl->current_fifo = &fifo;
+            found_anything = true;
+        }
+    }
+    return found_anything;
+}
+
+void Scheduler::ChangePriority(s32 channel_id, u32 new_priority) {
+    std::unique_lock lk(impl->scheduling_guard);
+    auto fifo_it = impl->channel_gpfifo_ids.find(channel_id);
+    if (fifo_it == impl->channel_gpfifo_ids.end()) {
+        return;
+    }
+    const size_t fifo_id = fifo_it->second;
+    auto& fifo = impl->gpfifos[fifo_id];
+    const auto old_priority = fifo.info->priority;
+    fifo.info->priority = new_priority;
+    impl->schedule_priority_queue.try_emplace(new_priority);
+    impl->schedule_priority_queue[new_priority].push_back(fifo_id);
+    impl->schedule_priority_queue[old_priority].remove_if(
+        [fifo_id](size_t id) { return id == fifo_id; });
+}
+
+void Scheduler::Yield() {
+    ASSERT(impl->current_fifo != nullptr);
+    impl->current_fifo->yield_count = impl->current_fifo->scheduled_count + 1;
+    Common::Fiber::YieldTo(impl->current_fifo->context, *impl->master_control);
+    gpu.BindChannel(impl->current_fifo->bind_id);
+}
+
+void Scheduler::CheckStatus() {
+    {
+        std::unique_lock lk(impl->scheduling_guard);
+        if (!impl->must_reschedule) {
+            return;
+        }
+    }
+    Common::Fiber::YieldTo(impl->current_fifo->context, *impl->master_control);
+    gpu.BindChannel(impl->current_fifo->bind_id);
+}
+
 void Scheduler::Push(s32 channel, CommandList&& entries) {
-    std::unique_lock lk(scheduling_guard);
-    auto it = channels.find(channel);
-    ASSERT(it != channels.end());
-    auto channel_state = it->second;
-    gpu.BindChannel(channel_state->bind_id);
-    channel_state->dma_pusher->Push(std::move(entries));
-    channel_state->dma_pusher->DispatchCalls();
+    std::unique_lock lk(impl->scheduling_guard);
+    auto it = impl->channel_gpfifo_ids.find(channel);
+    ASSERT(it != impl->channel_gpfifo_ids.end());
+    auto gpfifo_id = it->second;
+    auto& fifo = impl->gpfifos[gpfifo_id];
+    {
+        std::scoped_lock lk2(fifo.guard);
+        fifo.pending_work.emplace_back(std::move(entries));
+    }
+    if (impl->current_fifo != nullptr && impl->current_fifo->info->priority < fifo.info->priority) {
+        impl->must_reschedule = true;
+    }
+}
```

Comment:
If there's no correctness issue to marking a scheduler update as required, it would be cleaner to just unconditionally do it.
```suggestion
    std::scoped_lock lk(impl->scheduling_guard);

    // Get and ensure we have this channel.
    auto it = impl->channel_gpfifo_ids.find(channel);
    ASSERT(it != impl->channel_gpfifo_ids.end());

    auto gpfifo_id = it->second;
    auto& fifo = impl->gpfifos[gpfifo_id];

    // Add the new new work to the channel.
    {
        std::scoped_lock lk2(fifo.guard);
        fifo.pending_work.emplace_back(std::move(entries));
    }

    // Mark scheduler update as required.
    impl->must_reschedule = true;
}
```